File: Problems/ConnectFour/Go/board.go
package main

type DiscColor int

const (
	Empty DiscColor = iota
	RED
	YELLOW
)

type Board struct {
	rows int
	cols int
	grid [][]DiscColor
}

func NewBoard(rows, cols int) *Board {
	grid := make([][]DiscColor, rows)
	for r := 0; r < rows; r++ {
		grid[r] = make([]DiscColor, cols)
	}

	return &Board{
		rows: rows,
		cols: cols,
		grid: grid,
	}
}

func (b *Board) canPlace(column int) bool {
	if column < 0 || column >= b.cols {
		return false
	}
	if b.grid[0][column] != Empty {
		return false
	}
	return true
}

func (b *Board) PlaceDisc(column int, color DiscColor) []int {
	if !b.canPlace(column) {
		return []int{-1, -1}
	}

	for row := 0; row < b.rows-1; row++ {
		if b.grid[row+1][column] != Empty {
			b.grid[row][column] = color
			return []int{row, column}
		}
	}

	return nil
}

func (b *Board) CheckWin(row, colum int, color DiscColor) bool {
	dirs := [][2]int{
		{0, 1},
		{1, 0},
		{1, 1},
		{1, -1},
	}

	for _, dir := range dirs {
		dr := dir[0]
		dc := dir[1]

		count := 0

		count += b.coundDirection(row, colum, dr, dc, color)
		count += b.coundDirection(row, colum, -dr, -dc, color)

		if count >= 4 {
			return true
		}
	}

	return false
}

func (b *Board) coundDirection(row, col, dr, dc int, color DiscColor) int {
	count := 0

	for {
		row += dr
		col += dc

		if !b.isValid(row, col) {
			break
		}

		if b.grid[row][col] != color {
			break
		}

		count++
	}

	return count
}

func (b *Board) isValid(row, col int) bool {
	return row >= 0 && col >= 0 && row < b.rows && col < b.cols
}

func (b *Board) IsFull() bool {
	count := b.rows * b.cols

	for i := 0; i < b.rows; i++ {
		for j := 0; j < b.cols; j++ {
			if b.grid[i][j] != Empty {
				count--
			}
		}
	}

	if count == 0 {
		return true
	}
	
	return false
}

File: Problems/ConnectFour/Java/Board.java

public class Board {
    private int rows;
    private int cols;
    private DiscColor[][] grid;

    public Board(int rows, int cols) {
        this.rows = rows;
        this.cols = cols;
        this.grid = new DiscColor[rows][cols];
        initializeGrid(grid);
    }

    private void initializeGrid(DiscColor[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;

        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                grid[i][j] = DiscColor.EMPTY;
            }
        }
    }

    private boolean canPlace(int column) {
        if(column < 0 || column >= cols) {
            return false;
        }
        if(grid[0][column] != DiscColor.EMPTY) {
            return false;
        }
        return true;
    }

    public int[] placeDic(int column, DiscColor color) {
        if(canPlace(column)) {
            return new int[]{-1, -1};
        }

        for(int row = 0; row < rows - 1; row++) {
            if(grid[row + 1][column] != DiscColor.EMPTY) {
                grid[row][column] = color;
                return new int[]{row, column};
            }
        }

        return new int[]{-1, -1};
    }

    public boolean checkWin(int row, int col, DiscColor color) {
        int[][] dirs = new int[][] {
            {0, 1},
            {1, 0},
            {1, 1},
            {1, -1},
        };

        for(int[] dir : dirs) {
            int dr = dir[0];
            int dc = dir[1];

            int count = 0;

            count += countDirection(row, col, dr, dc, color);
            count += countDirection(row, col, -dr, -dr, color);

            if(count >= 4) {
                return true;
            }
        }

        return false;
    }

    public boolean isFull() {
        int count = rows * cols;
        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                if(grid[i][j] != DiscColor.EMPTY) {
                    count--;
                }
            }
        }
        return count == 0 ? true : false;
    }

    private int countDirection(int row, int col, int dr, int dc, DiscColor color) {
        int count = 0;
        while(true) {
            row += dr;
            count += dc;

            if(!isValid(row, col)) {
                break;
            }

            if(grid[row][col] != color) {
                break;
            }

            count++;
        }
        return count;
    }

    public boolean isValid(int row, int col) {
        return row >= 0 && col >= 0 && row < rows && col < cols;
    }
}